Issue/6244 Swipe to refresh

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6244 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
Adds swipe to refresh 

### Testing instructions
- Open more menu -> Inbox 
- Swipe to refresh. 
- Most likely the list won't change as it would be the same list. To properly test this open `wc-core` and dismiss some notes from there. Then swipe to refresh from Inbox. 

### Images/gif

https://user-images.githubusercontent.com/2663464/163158707-ec4392ee-af2f-41bc-8595-7976d4c0b146.mp4


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Add swipe to refresh for Inbox screen

JorgeMucientes committed:
Add test for swipe to refresh

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/inbox/InboxViewModel.kt:
```diff
@@ -62,11 +62,23 @@ class InboxViewModel @Inject constructor(
+                    isLoading = false,
+                    notes = notes,
+                    onRefresh = ::refreshNotes,
+                    isRefreshing = false
```
Hiding the refresh indicator as a reaction to the DB observer is fragile, it depends on the logic of the DB on re-emitting items without being changed, which means if anyone went to the Repo or the Store and added a `distinctUntilChanged`, it would cause the refresh indicator to stay always active:

https://user-images.githubusercontent.com/1657201/163587679-f9e66da6-65a9-449e-aa6f-c649fb33ea0e.mov

IMO, this logic should be moved to `refreshNotes`, with something like:
```kotlin
    private fun refreshNotes() {
        viewModelScope.launch {
            _inboxState.value = _inboxState.value?.copy(isRefreshing = true)
            inboxRepository.fetchInboxNotes()
           _inboxState.value = _inboxState.value?.copy(isRefreshing = false)
        }
    }
```

hichamboushaba reviewed (changes requested):
The code looks good, I left one comment about the logic for marking the refresh as completed.
And also, an important point, the app crashed for me when I tested it:
```    java.lang.IllegalArgumentException: No enum constant com.woocommerce.android.ui.inbox.domain.InboxNote.NoteType.WARNING
        at java.lang.Enum.valueOf(Enum.java:257)
        at com.woocommerce.android.ui.inbox.domain.InboxNote$NoteType.valueOf(Unknown Source:2)
        at com.woocommerce.android.ui.inbox.domain.InboxRepository.inboxNoteTypeFromString(InboxRepository.kt:75)
        at com.woocommerce.android.ui.inbox.domain.InboxRepository.toInboxNote(InboxRepository.kt:69)
        at com.woocommerce.android.ui.inbox.domain.InboxRepository.access$toInboxNote(InboxRepository.kt:15)
        at com.woocommerce.android.ui.inbox.domain.InboxRepository$observeInboxNotes$$inlined$map$1$2.emit(Collect.kt:137)
```
I workaround the error by adding a `Warning` type to `NoteType` enum, but I feel like this shouldn't crash if we get a type that we don't know and handle.

JorgeMucientes committed:
Ensure isRefreshing state is updated after fetching new notes

JorgeMucientes committed:
Handle exception for mapping unknown inbox note values to enum

JorgeMucientes committed:
Replace unneeded UPDATE type with missing type WARNING

JorgeMucientes committed:
Fix swipe to refresh unit test

JorgeMucientes commented:
Thanks for testing and looking into this @hichamboushaba. 

Entirely agree with both suggested changes, good points. In fact, i was able to experience the bug while testing another branch where the swipe to refresh spinner will stay there forever. If you swipe to refresh in airplane mode for example. 

As for the string to enum conversion. I agree, although I had some doubts about it first. In any case, I have added some error handling to catch the exception and set a default value. 
To give some context on this. The request in FluxC determines the default types we are expecting from the API in `WcInboxCore`: 
```kotlin 
        val INBOX_NOTE_TYPES_FOR_APPS = arrayOf(
            "info",
            "survey",
            "marketing",
            "warning"
        )
```
If there is a misalignment like there was in this PR then a crash would happen when converting from string to enum. My initial idea was to follow a "Crash early" approach instead of hiding the issue by catching the exception, as we shouldn't have unexpected values there. That said I realized, that right now we only care about survey type notes so we can safely fallback to `INFO` type if an unrecognized string value comes from API. The rest of the types don't have any other difference in the UI. In this case, I think it is safer to catch the exception and rely on a default value.

JorgeMucientes commented:
Thanks for testing and looking into this @hichamboushaba. 

Entirely agree with both suggested changes, good points. In fact, i was able to experience the bug while testing another branch where the swipe to refresh spinner will stay there forever. If you swipe to refresh in airplane mode for example. 

As for the string to enum conversion. I agree, although I had some doubts about it first. In any case, I have added some error handling to catch the exception and set a default value. 
To give some context on this. The request in FluxC determines the default types we are expecting from the API in `WcInboxCore`: 
```kotlin 
        val INBOX_NOTE_TYPES_FOR_APPS = arrayOf(
            "info",
            "survey",
            "marketing",
            "warning"
        )
```
If there is a misalignment like there was in this PR then a crash would happen when converting from string to enum. My initial idea was to follow a "Crash early" approach instead of hiding the issue by catching the exception, as we shouldn't have unexpected values there. That said I realized, that right now we only care about survey type notes. As the rest don't have any other difference in the UI. In this case, I think it is safer to catch the exception and rely on a default value.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/inbox/domain/InboxRepository.kt:
```diff
@@ -72,7 +72,9 @@ class InboxRepository @Inject constructor(
         when (typeName) {
             null -> NoteType.INFO
-            else -> enumValueOf(typeName.uppercase())
+            else -> runCatching {
+                enumValueOf(typeName.uppercase()) as NoteType
+            }.getOrDefault(NoteType.INFO)
```
💯

hichamboushaba reviewed (approved):
Nice work, thanks for the fixes @JorgeMucientes.

kidinov commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/inbox/InboxViewModelTest.kt:
```diff
@@ -71,6 +78,22 @@ class InboxViewModelTest : BaseUnitTest() {
         }
 
+    @Test
+    fun `Given inbox notes loaded, when refresh notes, show is refreshing state and fetch notes`() =
```
nit: should start with a lower case

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/inbox/Inbox.kt:
```diff
@@ -58,7 +61,11 @@ fun Inbox(state: InboxState) {
     when {
         state.isLoading -> InboxSkeleton()
         state.notes.isEmpty() -> InboxEmptyCase()
```
Shouldn't an empty case also be wrapped in a swipe to refresh?

JorgeMucientes replied:
That is a good point @kidinov. Although in this case I don't think is needed. Every time you enter Inbox we fetch the notes from the API and refresh the list with the results. So if the user sees the empty case it is very unlikely that refreshing again through "swipe-to_refresh" is going to display anything different.

kidinov replied:
@JorgeMucientes So when there is something in the DB we show local data without hitting the API? If it is so then it doesn't sound right to me, I think we should show cached data but still do the call in the background and update the list when it comes back

btw, feel free to merge this, I am not blocking anything, just talking =))

JorgeMucientes replied:
No, I might not have explained myself right. The behavior is as follows @kidinov 
When opening Inbox:
- We display data from DB if there is any. If there is nothing then a skeleton is shown while we fetch data from API. 
- We fetch Inbox notes from the API **always** when entering the screen. Wheter there is data in DB or not. 
- When the response from API is back, if the response is successful, we will refresh the UI with the latest data from the API. If the request fails for any reason, we will display data from DB or the empty case with an error feedback.

So with that behavior in mind. If the user is seeing an empty case after entering Inbox screen, that is because the API has returned an empty list of notes or the request is failing (due to connectivity, server issues or whatever). In both cases, refreshing the list right away from swipe to refresh doesn't make sense to me, as the results are going to be the same: empty. 

That said, nesting the empty case inside the swipe to refresh is trivial, so if I'm missing something let me know. I'm totally open to enable swipe to refresh in the empty case.

kidinov replied:
@JorgeMucientes 
I'm not following 😢. What is the dogmatical difference between an empty and not empty state? Aren't they both children of the "success" screen state, the first is when items count is 0 and the second is when it's > 0?

Let say
* I have 1 item and a new item appeared on the backed: I have a way to see this with a swipe to refresh
* I have 0 items and a new item appeared on the backed: I don't have a way to see the new item with a swipe to refresh and have to leave the screen and enter again 

I think having a swipe to refresh on an empty states it's kind of generic practise

JorgeMucientes replied:
> What is the dogmatical difference between an empty and not empty state? 

Yeah that is fair point. There really is no difference. I was looking at it more from the point of how useful it'd be for the user to have swipe to refresh on empty case given how the screen currently works: refreshing the notes always when entering. And also considering that notes don't change frequently. That said, this lack of usefulness also applies to swipe to refresh when there are some notes. 

> I think having a swipe to refresh on an empty states it's kind of generic practise

That is also true. And  I'd say that reason that most convinces me 🤣 . Added it! 

 Thanks a lot for sharing your thoughts @kidinov

JorgeMucientes committed:
Minor code style fix

JorgeMucientes committed:
Add swipe to refresh to empty case

JorgeMucientes committed:
Merge branch 'issue/5876-dismiss-all-notes' into issue/6244-swipe-to-refresh

JorgeMucientes commented:
This is ready for merging as the parent branch was merged and this is now updated and bases from`trunk`. 

Let me know if there is any feedback left to add before I merge 🙂 I added the missing swipe to refresh to the empty case.
